feat(ssh_tunnel): SQLAlchemy Form UI

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
Add our a new `SSHTunnelForm` component that will be used to set tunnels for our Databases, right now only works when using SQLAlchemy URIs.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
1. Make sure the SSH_TUNNELING feature flag is enabled
2. open the create database modal
3. click on SQL Alchemy link so the SQLA form is displayed
4. Enable the new SSH Tunnel switch
5. Make use of the form so you can connect a DB with a SSH Tunnel attached to it

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [X] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

hughhhh committed:
save

hughhhh committed:
create migration

hughhhh committed:
created schema and rename

hughhhh committed:
linting

hughhhh committed:
fix encrpytions

hughhhh committed:
remove map tabl

hughhhh committed:
fix linting

hughhhh committed:
add constraint

hughhhh committed:
add fk to migration

hughhhh committed:
init

hughhhh committed:
update all the examples

hughhhh committed:
change remaining bits

hughhhh committed:
add id

hughhhh committed:
use factory instead

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
Merge branch 'master' into ref-get-sqla-engine-2

hughhhh committed:
fix confict

hughhhh committed:
fix conflict

hughhhh committed:
setup return value for contextmanager

hughhhh committed:
add sshtunnel pip

hughhhh committed:
updates test

hughhhh committed:
fix linting

hughhhh committed:
renaming function

hughhhh committed:
fix test

hughhhh committed:
Merge branch 'ref-get-sqla-engine-2' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
add schema to test_connection api

hughhhh committed:
fix get engine to return contextmanager

hughhhh committed:
why

hughhhh committed:
yerp

hughhhh committed:
update typing

hughhhh committed:
update comment

hughhhh committed:
Merge branch 'ref-get-sqla-engine-2' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
save

hughhhh committed:
save

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into ref-get-sqla-engine-2

hughhhh committed:
fix pylint

hughhhh committed:
Merge branch 'ref-get-sqla-engine-2' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
last one

hughhhh committed:
Merge branch 'ref-get-sqla-engine-2' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
update naming on ssh tunnel

hughhhh committed:
Merge branch 'master' into ref-get-sqla-engine-2

hughhhh committed:
Merge branch 'ref-get-sqla-engine-2' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
fix renaming

hughhhh committed:
fix renaming 2

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
oops

hughhhh committed:
fix linting errors

Antonio-RiveroMartnez committed:
feat(ssh_tunnel): DAO Changes for SSH Tunnel (#22120)

hughhhh committed:
fix merge conflicts

hughhhh committed:
Merge branch 'create-sshtunnelconfig-tbl' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

Antonio-RiveroMartnez committed:
feat(ssh_tunnel): Delete command & exceptions (#22131)

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
Merge branch 'create-sshtunnelconfig-tbl' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
fix indenting for superset/databases/commands/validate.py

hughhhh committed:
change tablename

Antonio-RiveroMartnez committed:
feat(ssh_tunnel): DELETE SSH Tunnels API (#22153)

hughhhh committed:
Revert "feat(ssh_tunnel): DELETE SSH Tunnels API" (#22156)

Antonio-RiveroMartnez committed:
feat(ssh_tunnel): Update command & exceptions (#22132)

hughhhh committed:
forgot server_port

hughhhh committed:
bind_port + bind_host :)

hughhhh committed:
oops

hughhhh committed:
fix linting

Antonio-RiveroMartnez committed:
feat(ssh_tunnel): SSH Tunnel updates from Code Review (#22182)

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
feat(ssh_tunnel): Create command & exceptions  (#22148)

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Initial DELETE and CREATE ssh tunnel APIs
- Add happy path tests for each API

NOTE:
- Updating a DB might also result in creating a Tunnel, that flow will get tested in the UPDATE API changes.

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Uncomment ssh tunnel creation for TestConnection command
- Update our tests
- Add update changes in API

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Adjust our tests so it's clear our endpoint uses PATCH and not PUT
- Add delete cascading test: When deleting a database, the tunnel should be deleted

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add not found test when deleting the tunnel

Antonio-RiveroMartnez committed:
SSH Tunnel:

- FIx linting errors

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix pre-commit errors

hughhhh committed:
Update schemas.py

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Remove bind properties from schema used in TestConnection
- Add test for SSH Tunnel failure and no DB creation

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Make server_address, server_port and username required fields for our SSHTunnel schema
- Update the tests so we can consider new message with required fields

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
chore(ssh-tunnel): create `contextmanager` for sql.inspect (#22251)

hughhhh committed:
fix lint

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Changes from Code review feedback
- Solve conflict with base branch

hughhhh committed:
fix migrations

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Remove safe decorator so SIP-41 errors can be returned

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix update tests

hughhhh committed:
Revert "chore(ssh-tunnel): create `contextmanager` for sql.inspect (#22251)"

This reverts commit 44487396cef334d57bab8fd97100d943a90e0934.

hughhhh committed:
debugging

hughhhh committed:
debugging

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Mask passwords for SSH when creating a new one
- Fix arg name in test connection

Antonio-RiveroMartnez committed:
fix(ssh_tunnel): Address Base PR comments from peer review (#22306)

hughhhh committed:
fix pre-commit

hughhhh committed:
working changes

hughhhh committed:
refactor bind_host and bind_port

hughhhh committed:
refactor bind_host and bind_port

hughhhh committed:
refactor create flow for temp ssh tunnels

hughhhh committed:
remove logger

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Use SSHTunnel commands instead of DAOs
- Update tests
- Fix linting and pre-commit

Antonio-RiveroMartnez committed:
chore(ssh_tunnel): Add extra tests to SSHTunnel commands (#22372)

hughhhh committed:
add flush to allow database.id to be populated

hughhhh committed:
Merge branch 'create-sshtunnelconfig-tbl' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Flush only if creating a ssh tunnel

hughhhh committed:
make sure to use inspector with context

hughhhh committed:
remove id and database_id

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Do not return passwords for SSH Tunnel in any API call

hughhhh committed:
uselist

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Using nested transactions so we can rollback if anything fails

hughhhh committed:
feat(ssh-tunnel): ssh manager config + feature flag (#22201)

hughhhh committed:
update kwarg function name

hughhhh committed:
chore(ssh-tunnel): Move SSHManager to extensions pattern (#22433)

hughhhh committed:
add flag to indicate ssh tunneling is enabled for this engine

hughhhh committed:
Update superset/migrations/versions/2022-10-20_10-48_f3c2d8ec8595_create_ssh_tunnel_credentials_tbl.py

Co-authored-by: Elizabeth Thompson <eschutho@gmail.com>

hughhhh committed:
Update superset/migrations/versions/2022-10-20_10-48_f3c2d8ec8595_create_ssh_tunnel_credentials_tbl.py

Co-authored-by: Elizabeth Thompson <eschutho@gmail.com>

hughhhh committed:
fix linting

Antonio-RiveroMartnez committed:
Merge branch 'create-sshtunnelconfig-tbl' into api_ssh_tunnel

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Extend GET endpoint on Database APi so it includes SSH Tunnel if any

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Initial SSH Tunnel UI: Adding SSH form to SQLA form

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Handle Database edition with SSH Tunnels
- Add masked password, private_key and private_key_password to GET call so when editing a DB we can use those and unmask before updating

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Remove ssh_tunnel form DB when toggling off the config switch (opting out when creating or editing a DB w/o a tunnel)

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add masked values for password, private_key and private_key_password for POST and PUT endpoints
- Update our tests

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix incompatible types pre-commit error

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Add new reducer action types tests

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

Antonio-RiveroMartnez committed:
Merge branch 'create-sshtunnelconfig-tbl' into api_ssh_tunnel

Antonio-RiveroMartnez committed:
Merge master into ssh_sqla_form

hughhhh committed:
fix requirements

Antonio-RiveroMartnez committed:
Merge branch 'create-sshtunnelconfig-tbl' into api_ssh_tunnel

Antonio-RiveroMartnez committed:
Merge branch 'api_ssh_tunnel' into ssh_sqla_form

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Use SSH Tunnel only if feature flag is enabled

hughhhh committed:
Merge branch 'master' of https://github.com/apache/superset into create-sshtunnelconfig-tbl

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add util method to mask password
- Adjust tests to consider password masking

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Remove the extra description on GET endpoint for databases

hughhhh committed:
get df with get_raw_connection function

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix pre-commit error

Antonio-RiveroMartnez committed:
feat(ssh_tunnel): APIs for SSH Tunnels (#22199)

Co-authored-by: hughhhh <hughmil3s@gmail.com>

Antonio-RiveroMartnez committed:
Merge branch 'api_ssh_tunnel' into ssh_sqla_form

Antonio-RiveroMartnez committed:
Merge branch 'create-sshtunnelconfig-tbl' into api_ssh_tunnel

Antonio-RiveroMartnez committed:
Merge branch 'api_ssh_tunnel' into ssh_sqla_form

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix pylint error

Antonio-RiveroMartnez committed:
Merge master

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Add visual test to confirm the new SSH Tunnel form fields are present if we click the toggle button

Antonio-RiveroMartnez committed:
Merge branch 'master' into ssh_sqla_form

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Add extra UI tests for our SSH Tunnel Form

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Add new test to consider changing the login method used in SSH Tunnel Form

Antonio-RiveroMartnez committed:
Merge branch 'master' into ssh_sqla_form

hughhhh commented on superset/databases/ssh_tunnel/dao.py:
```diff
@@ -24,3 +26,27 @@
+        The masked values should be unmasked before the ssh tunnel is updated.
+        """
+        if properties.get("password") == PASSWORD_MASK:
+            properties["password"] = model.password
```
we should move this logic to the frontend, basically if the field hasn't been updated we shouldn't send it in the payload

Antonio-RiveroMartnez replied:
The response of the GET already has the masked password thus it's stored in the reducer, so when calling the API to update let's say the `server_port`, we send the whole `ssh_tunnel` object, not just updated properties. Because of that, the API would assume you're trying to update the password when you were not. Pretty much like we are doing for other masked values in our forms.

hughhhh commented on superset-frontend/src/views/CRUD/data/database/DatabaseModal/index.tsx:
```diff
@@ -1428,7 +1486,36 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
                 testInProgress={testInProgress}
-              />
+              >
+                <SSHTunnelForm
```
@yousoph do we want to show the ssh tunnel switch even though tunneling is turned off?

Antonio-RiveroMartnez committed:
Merge branch 'master' into ssh_sqla_form

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Hide the entire form behind the feature flag

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Fix TestConnection when editing a DB with SSH Tunnel attached to it
- Add new unmask_password_info util method for SSH Tunnels

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Add ssh_tunnel id in our response for the GET endpoint of DB so we can call test_connection endpoint without depending on DB's name

hughhhh commented on superset-frontend/src/views/CRUD/data/database/DatabaseModal/DatabaseConnectionForm/index.tsx:
```diff
@@ -73,6 +73,7 @@ export interface FieldPropTypes {
     onRemoveTableCatalog: (idx: number) => void;
```
also we need to check the available endpoint to see if a given engine allows ssh tunneling
https://github.com/apache/superset/blob/master/superset/db_engine_specs/base.py#L196

Antonio-RiveroMartnez replied:
Oh! Ok, yeah, IIRC there a `engine_information` property we are getting in the UI with the `supports_file_upload`, so we might need to include this one there and use it. Thanks!

Antonio-RiveroMartnez committed:
Merge branch 'master' into ssh_sqla_form

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Use the allow_ssh_tunneling flag from the db_engine along with the feature flag to display the SSH Tunnel form, so, if a DB engine does not support SSH Tunnels, nothing gets displayed. Right now, only Postgres does

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Fix tests with proper flag values

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- More fixes to tests

hughhhh reviewed (approved):
LGTM

Antonio-RiveroMartnez committed:
SSH Tunnel UI:

- Fix test
